File: leetcode/0355. Design Twitter/Solution.swift
// https://leetcode.com/problems/design-twitter/

class Twitter {

    private struct Following: Hashable {
        let followerId: Int
        let followeeId: Int
    }

    private struct Tweet {
        let index: Int
        let userId: Int
        let tweetId: Int
    }

    private let feedMaxLength = 10

    private var followingSet = Set<Following>()

    private var tweets = Array<Tweet>()
    private var tweetCounter = Int(0)

    init() {}

    func postTweet(_ userId: Int, _ tweetId: Int) {
        tweets.append(Tweet(index: tweetCounter, userId: userId, tweetId: tweetId))
        tweetCounter += 1
    }

    func getNewsFeed(_ userId: Int) -> [Int] {
        let feedIds = [userId] + followingSet.filter { $0.followerId == userId }.map { $0.followeeId }
        let allTweetIds = tweets.filter { feedIds.contains($0.userId) }.sorted { $0.index > $1.index }.map { $0.tweetId }
        let returnCount = allTweetIds.count > feedMaxLength ? feedMaxLength : allTweetIds.count

        return returnCount < 1 ? [] : Array(allTweetIds[0..<returnCount])
    }

    func follow(_ followerId: Int, _ followeeId: Int) {
        followingSet.insert(Following(followerId: followerId, followeeId: followeeId))
    }

    func unfollow(_ followerId: Int, _ followeeId: Int) {
        followingSet.remove(Following(followerId: followerId, followeeId: followeeId))
    }
}

File: leetcode/0355. Design Twitter/Solution.py
# https://leetcode.com/problems/design-twitter/

from typing import List
from collections import defaultdict

class Twitter:
    def __init__(self):
        self.followed_users = defaultdict(set) # userId: (followeeId)
        self.tweets = defaultdict(list) # userId: [tweetId]
        
        self.tweet_count = 1
        self.tweet_order = dict() # tweetId: tweet_count

    def postTweet(self, userId: int, tweetId: int) -> None:
        self.tweets[userId].append(tweetId)
        
        self.tweet_order[tweetId] = self.tweet_count
        self.tweet_count += 1

    def getNewsFeed(self, userId: int) -> List[int]:
        self.followed_users[userId].add(userId)
        
        tweets = []
        for f in self.followed_users[userId]:
            tweets += self.tweets[f]
        
        return sorted(tweets, key=lambda x: -self.tweet_order[x])[:10]

    def follow(self, followerId: int, followeeId: int) -> None:
        self.followed_users[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followeeId in self.followed_users[followerId]:
            self.followed_users[followerId].remove(followeeId)

twitter = Twitter()
twitter.postTweet(1, 5)
assert twitter.getNewsFeed(1) == [5]
twitter.follow(1, 2)
twitter.postTweet(2, 6)
assert twitter.getNewsFeed(1) == [6, 5]
twitter.unfollow(1, 2)
assert twitter.getNewsFeed(1) == [5]
